### PR TITLE
Feature/trainer kwargs

### PR DESCRIPTION
**Pull request by teodor-r**
Added `get_val_mask_func_kwargs` and  `get_trainer_func_kwargs` arguments for Transformer-based models to allow keyword arguments in custom functions used for model training

### CODE REVIEW

File: rectools/models/nn/transformers/bert4rec.py [hunk 1]
```diff
@@ -88,6 +88,7 @@ def __init__(
         mask_prob: float = 0.15,
         shuffle_train: bool = True,
         get_val_mask_func: tp.Optional[ValMaskCallable] = None,
+        get_val_mask_func_kwargs: tp.Optional[InitKwargs] = None,
```
Comment (blondered) [inline rectools/models/nn/transformers/bert4rec.py:91]: Please update docstrings

File: tests/models/nn/transformers/test_sasrec.py [hunk 2]
```diff
@@ -968,6 +968,8 @@ def initial_config(self) -> tp.Dict[str, tp.Any]:
             "backbone_type": TransformerTorchBackbone,
             "get_val_mask_func": leave_one_out_mask,
             "get_trainer_func": None,
+            "get_val_mask_func_kwargs": None,
```
Comment (blondered) [inline tests/models/nn/transformers/test_sasrec.py:971]: Please add tests for custom trainer func with kwargs and custom validation func with kwargs

File: tests/models/nn/transformers/test_bert4rec.py [hunk 3]
```diff
@@ -310,14 +322,29 @@ def get_trainer() -> Trainer:
         ),
     )
     @pytest.mark.parametrize("u2i_dist", ("dot", "cosine"))
+    @pytest.mark.parametrize(
```
Comment (blondered) [inline tests/models/nn/transformers/test_bert4rec.py:325]: Parametrize doesn't work when we have only one variant in a tuple

File: tests/models/nn/transformers/test_bert4rec.py [hunk 4]
```diff
@@ -114,6 +118,14 @@ def get_trainer() -> Trainer:
 
         return get_trainer
 
+    @pytest.fixture
+    def factory_get_trainer_func(self) -> TrainerCallable:
+        def get_trainer_with_kwargs(**get_trainer_kwargs) -> Trainer:
+            get_trainer_kwargs = get_trainer_kwargs or {}
+            return Trainer(**get_trainer_kwargs)
```
Comment (blondered) [inline tests/models/nn/transformers/test_bert4rec.py:125]: Let's not generate Trainer from passed kwargs but instead test get_trainer_func that accepts kwargs

File: rectools/models/nn/transformers/base.py [hunk 5]
```diff
@@ -161,7 +161,7 @@ def _serialize_type_sequence(obj: tp.Sequence[tp.Type]) -> tp.Tuple[str, ...]:
 ]
 
 
-ValMaskCallable = Callable[[], np.ndarray]
```
Comment (blondered) [inline rectools/models/nn/transformers/base.py:164]: Please fix linters

Please update changelog

File: rectools/models/nn/transformers/bert4rec.py [hunk 6]
```diff
@@ -301,6 +304,10 @@ class BERT4RecModel(TransformerModelBase[BERT4RecModelConfig]):
         When set to ``None``, "cuda" will be used if it is available, "cpu" otherwise.
         If you want to change this parameter after model is initialized,
         you can manually assign new value to model `recommend_torch_device` attribute.
+    get_val_mask_func_kwargs: optional(InitKwargs), default ``None``
```
Comment (blondered) [inline rectools/models/nn/transformers/bert4rec.py:307]: We need to write everywhere in docstrings "Make sure all dict values have JSON serializable types."
Otherwise we will fail to serialize the model

File: tests/models/nn/transformers/test_bert4rec.py [hunk 7]
```diff
@@ -114,6 +123,38 @@ def get_trainer() -> Trainer:
 
         return get_trainer
 
+    @pytest.fixture
+    def get_custom_trainer_func(self) -> TrainerCallable:
+        def get_trainer_func(**kwargs: Unpack[KwargsSpec]) -> Trainer:
```
Comment (blondered) [inline tests/models/nn/transformers/test_bert4rec.py:128]: Since this is a test, we should write it the way user would use it.
And in this case it should be `def get_trainer_func(max_epochs: int,  accelerator: str) -> Trainer:`
Because here he wants hit triner func to change variants of max_epochs and accelerator and accept them from model config.

File: tests/models/nn/transformers/test_bert4rec.py [hunk 8]
```diff
@@ -114,6 +123,38 @@ def get_trainer() -> Trainer:
...
+                accelerator=accelerator,
+                enable_checkpointing=False,
+                devices=1,
+            )
+
+        return get_trainer_func
+
+    @pytest.fixture
+    def get_custom_val_mask_func(self) -> ValMaskCallable:
+        def get_val_mask_func(interactions: pd.DataFrame, **kwargs: InitKwargs) -> np.ndarray:
```
Comment (blondered) [inline tests/models/nn/transformers/test_bert4rec.py:145]:
```suggestion
        def get_val_mask_func(interactions: pd.DataFrame, val_users: tp.List[int]) -> np.ndarray:
```
(see comment for trainer func)

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Uncommited
+
+### Added 
+- `get_val_mask_func_kwargs` and  `get_trainer_func_kwargs` kwargs for corresponding functions
```
Comment (blondered) [inline CHANGELOG.md:11]:
```suggestion
- `get_val_mask_func_kwargs` and  `get_trainer_func_kwargs` arguments for Transformer-based models to allow keyword arguments in custom functions used for model training.
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Uncommited
```
Comment (blondered) [inline CHANGELOG.md:8]: Linters are failing (isort)

File: tests/models/nn/transformers/test_bert4rec.py [hunk 11]
```diff
@@ -816,6 +883,64 @@ def test_get_dataloader_val(
...
+        "val_batch, val_users",
+        (
+            (
+                {
+                    "x": torch.tensor([[0, 2, 4, 1]]),
+                    "y": torch.tensor([[3]]),
+                    "yw": torch.tensor([[1.0]]),
+                    "negatives": torch.tensor([[[5, 2]]]),
+                },
+                [10, 30],
```
Comment (blondered) [inline tests/models/nn/transformers/test_bert4rec.py:896]: val batch doesn't differ. Can we find a case where it differs to make sure all works as expected?

File: tests/models/nn/transformers/test_bert4rec.py [hunk 12]
```diff
@@ -816,6 +883,64 @@ def test_get_dataloader_val(
...
+        ),
+    )
+    def test_get_dataloader_val_with_kwargs(
+        self,
+        dataset: Dataset,
+        val_batch: tp.Dict[tp.Any, tp.Any],
+        val_users: tp.List,
+    ) -> None:
+
+        def get_custom_val_mask_func(interactions: pd.DataFrame, **kwargs: tp.Dict[str, tp.List]) -> np.ndarray:
```
Comment (blondered) [inline tests/models/nn/transformers/test_bert4rec.py:916]:
```suggestion
        def get_custom_val_mask_func(interactions: pd.DataFrame, val_users: tp.List[int]) -> np.ndarray:
```

File: rectools/models/nn/transformers/data_preparator.py [hunk 13]
```diff
@@ -84,6 +86,9 @@ def from_interactions(
         return cls(sessions=sessions, weights=weights)
 
 
+# pylint: disable=too-many-instance-attributes
```
Comment (feldlime) [inline rectools/models/nn/transformers/data_preparator.py:89]: Now this applied to the whole file below. Let's please specify it only for classes where we really need it

File: rectools/models/nn/transformers/data_preparator.py [hunk 14]
```diff
@@ -150,6 +160,13 @@ def get_known_item_ids(self) -> np.ndarray:
         """Return external item ids from processed dataset in sorted order."""
         return self.item_id_map.get_external_sorted_by_internal()[self.n_item_extra_tokens :]
 
+    @staticmethod
+    def _get_kwargs(actual_kwargs: tp.Optional[InitKwargs]) -> InitKwargs:
```
Comment (feldlime) [inline rectools/models/nn/transformers/data_preparator.py:164]: The name currently doesn't really reflect the meaning. Please rename it to `normalize_kwargs` or `ensure_kwargs_dict` or smth like this. I prefer the 2nd option since it's the most clear one

File: tests/models/nn/transformers/test_bert4rec.py (hunk 11 above)
Comment (teodor-r) [reply tests/models/nn/transformers/test_bert4rec.py:896]: A user with id 30 is only suitable for the model. All but 10 after clipping do not fit the length of the story, 10 does not fit because the target is a cold item

File: rectools/models/nn/transformers/data_preparator.py (hunk 14 above)
Comment (teodor-r) [reply rectools/models/nn/transformers/data_preparator.py]: fixed

File: rectools/models/nn/transformers/data_preparator.py (hunk 13 above)
Comment (teodor-r) [reply rectools/models/nn/transformers/data_preparator.py]: fixed

File: tests/models/nn/transformers/test_bert4rec.py (hunk 12 above)
Comment (teodor-r) [reply tests/models/nn/transformers/test_bert4rec.py]: fixed

File: CHANGELOG.md (hunk 10 above)
Comment (teodor-r) [reply CHANGELOG.md]: fixed

File: CHANGELOG.md (hunk 9 above)
Comment (teodor-r) [reply CHANGELOG.md]: edited

**Review by blondered — APPROVED** [review]
